File: WeeklyCode/week5/Construct Binary Tree from Preorder and Inorder Traversal/chenhao.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param preorder, a list of integers
    # @param inorder, a list of integers
    # @return a tree node
    def buildTree(self, preorder, inorder):
        return self.build(preorder,inorder,0,0,len(preorder))
        
    def build(self,preorder,inorder,preOffset,inOffset,length):
        if length<=0:
            return None
        if length==1:
            return TreeNode(preorder[preOffset])
        
        rootValue=preorder[preOffset]
        inIndex=inorder.index(rootValue,inOffset,inOffset+length)
        leftLength=inIndex-inOffset
        left=self.build(preorder,inorder,preOffset+1,inOffset,leftLength)
        right=self.build(preorder,inorder,preOffset+1+leftLength,inIndex+1,length-1-leftLength)
        
        root=TreeNode(rootValue)
        root.left=left
        root.right=right
        
        return root

File: WeeklyCode/week11/Construct Binary Tree from Preorder and Inorder Traversal/chenhao.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    int[] pre,in;
    
    private TreeNode build(int p,int i,int len){
        if(len<=0){
            return null;
        }
        TreeNode root=new TreeNode(pre[p]);
        int leftLen=0;
        while(in[i+leftLen]!=pre[p]){
            leftLen++;
        }
        root.left=build(p+1,i,leftLen);
        int rightLen=len-1-leftLen;
        root.right=build(p+1+leftLen,i+1+leftLen,rightLen);
        return root;
    }
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.pre=preorder;
        this.in=inorder;
        return build(0,0,preorder.length);
    }
}